### PR TITLE
Allow specifying custom directory to find appium.txt

### PR DESCRIPTION
**Pull request by mlainez**
Hi,

I'm trying to allow the use of a custom directory for your appium.txt files. I don't mind that the files should be named appium.txt, I just want to be able to pick a specific location where arc is supposed to find it.

Let me know what you think.

### CODE REVIEW

File: lib/appium_console.rb [hunk 1]
```diff
@@ -9,9 +9,17 @@ def self.warn args
 
 require 'pry'
 
-Pry.send(:define_singleton_method, :pry_load_appium_txt) do |opts={}|
-  verbose = opts.fetch :verbose, false
-  Appium.load_appium_txt file: Dir.pwd + '/appium.txt', verbose: verbose
+Pry.send(:define_singleton_method, :pry_load_appium_txt) do |opts = {}|
+  if ARGV.any?
```
Comment (bootstraponline) [inline lib/appium_console.rb]: `ARGV.any?` doesn't ensure we have at least two elements (`ARGV[0]` & `ARGV[1]`). I think `ARGV.length >= 2` would be better.

Comment (mlainez) [inline lib/appium_console.rb]: You're right. I'll update it.

File: lib/appium_console.rb [hunk 2]
```diff
@@ -9,9 +9,17 @@ def self.warn args
...
+Pry.send(:define_singleton_method, :pry_load_appium_txt) do |opts = {}|
+  if ARGV.size >= 2
+    command       = ARGV[0].to_sym
+    option        = ARGV[1]
+    opts[command] = option
+  end
+
+  verbose       = opts.fetch :verbose, false
+  dir           = opts.fetch :appium_dir, Dir.pwd
+  path          = dir + "/appium.txt"
```
Comment (bootstraponline) [inline lib/appium_console.rb]: This should be `File.join(dir, 'appium.txt')` which uses `File::SEPARATOR`

Comment (mlainez) [inline lib/appium_console.rb]: Indeed, let me make that quick change.

File: lib/appium_console.rb [hunk 3]
```diff
@@ -9,9 +9,17 @@ def self.warn args
...
+Pry.send(:define_singleton_method, :pry_load_appium_txt) do |opts = {}|
+  if ARGV.size >= 2
+    command       = ARGV[0].to_sym
+    option        = ARGV[1]
+    opts[command] = option
+  end
+
+  verbose       = opts.fetch :verbose, false
+  dir           = opts.fetch :appium_dir, Dir.pwd
+  path          = File.join dir, "/appium.txt"
```
Comment (bootstraponline) [inline lib/appium_console.rb]: the `/` needs to be deleted

Comment (mlainez) [inline lib/appium_console.rb]: My mistake... Overall, I must say I'm not really convinced about the current arg management stuff. I saw issue #43 and I might be able to give it a try next week and add tests to it so this pr will become irrelevant. I'd rather do something clean and refactor stuff along the way than have this "thing" merged. I'll keep you posted.

Comment (bootstraponline) [inline lib/appium_console.rb]: Sounds good.

**Comment by mlainez** [comment]
@bootstraponline took me longer that I thought to find the time but I made some first pass at a refactoring, also adding the ability for speciiying a toml file anywhere at execution. Let me know what you think and what should be improved.

File: appium_console.gemspec [hunk 4]
```diff
@@ -28,6 +28,7 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency 'pry', '~> 0.10.1'
   s.add_runtime_dependency 'bond', '~> 0.5.1'
   s.add_runtime_dependency 'spec', '>= 5.3.4'
+  s.add_runtime_dependency 'thor', '>= 0.19'
```
Comment (bootstraponline) [inline appium_console.gemspec]: why do we need thor at runtime?

File: lib/appium_console.rb [hunk 5]
```diff
@@ -7,45 +9,47 @@ def self.warn args
...
 
-require 'pry'
-
-Pry.send(:define_singleton_method, :pry_load_appium_txt) do |opts={}|
-  verbose = opts.fetch :verbose, false
-  Appium.load_appium_txt file: Dir.pwd + '/appium.txt', verbose: verbose
-end
+module Appium
+  module Console
+    def self.setup appium_txt_path
```
Comment (bootstraponline) [inline lib/appium_console.rb]: In general the appium Ruby code prefers `class << self` over `def self.foo`

File: appium_console.gemspec (hunk 4 above)
Comment (bootstraponline) [inline appium_console.gemspec]: ignore this comment, it's for the CLI features.

**Comment by bootstraponline** [comment]
The code looks awesome! Once the style issue is fixed then this can be merged.

I'm assuming you've tried the code and it works? Unfortunately the ruby console doesn't have tests.

**Comment by mlainez** [comment]
I did manually test it indeed but the upgrade somehow did not work with my setup, actually the upgrade from the original version does not work either... Maybe you can give it a try? 

I'm not confortable either that there are no tests. However, although I have some experience with testing, I have no clue how to properly test a cli... I know the aruba gem that is supposed to help you with this but the question is more what to test and how to test it... That's where I'm a bit lost, any suggestions would be appreciated and I could try to add some acceptance tests for each available command.

**Comment by bootstraponline** [comment]
> I did manually test it indeed but the upgrade somehow did not work with my setup, actually the upgrade from the original version does not work either... Maybe you can give it a try?

I think we should just delete the upgrade command. People can upgrade the gem like all the other gems (via rubygems).

**Comment by bootstraponline** [comment]
> I have no clue how to properly test a cli... I know the aruba gem that is supposed to help you with this but the question is more what to test and how to test it... That's where I'm a bit lost, any suggestions would be appreciated and I could try to add some acceptance tests for each available command.

Testing thor based gems is easy. My approach would be to write an rspec test that invokes `bin/arc` via an absolute path in a subprocess, and then make assertions on the console output. This will allow end to end testing of all the supported arguments.

**Comment by mlainez** [comment]
Got rid of upgrade, I was about to suggest the same thing. Style is changed as well.

I took the liberty of adding the toml command because I'm using appium with cucumber and I'm getting a bit bored of changing my appium.txt app path each time I use arc then execute tests.

**Comment by bootstraponline** [comment]
> I took the liberty of adding the toml command because I'm using appium with cucumber and I'm getting a bit bored of changing my appium.txt app path each time I use arc then execute tests.

:+1: 

Thanks for improving the gem. The code looks good to me so I'll merge this now. E2E tests could be a new PR if you have time.

**Comment by mlainez** [comment]
I'll look into it, I'm interested in the topic. Beware that upgrade is no longer there so if you bump the version you might want to go to 2.0

**Comment by bootstraponline** [comment]
> I'll look into it, I'm interested in the topic.

Here's an [example test](https://github.com/grosser/parallel_tests/blob/master/spec/integration_spec.rb#L41) that asserts on process output.

> Beware that upgrade is no longer there so if you bump the version you might want to go to 2.0

Agreed

**Comment by bootstraponline** [comment]
[2.0 has been released to rubygems](https://rubygems.org/gems/appium_console) and I added you to the ruby team.

**Comment by mlainez** [comment]
Thanks, updating my projects right now.

**Pull request merged**